index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
```

Comment:
Nice way but it is becoming quite complicated, isn't it? 

This is not the best way in my opinion:

```
type Dispatch<A, O> = (action: A) => O;
```

…but instead you can try this:

```
type Dispatch = <A, O>(action: A) => O;
```

Also I've checked source code ( see: https://github.com/reactjs/redux/blob/master/src/createStore.js#L143 ) - `.dispatch()` returns reference to action passed in so it actually is:

```
type Dispatch = <A>(action: A) => A;
```

And then :

```
interface Store<S> {
    dispatch: Dispatch;
}
```

So when anyone uses his store:

```
myStore.dispatch({anything})  // ok

// and if we have variable 'x' of type 'XType'
const done = myStore.dispatch(x); // ok
// TS compiler will infer 'done'  type as 'XType'


const newAction: MyAction = …;
myStore.dispatch<MyAction>(newAction);  // ok
const doneAction = myStore.dispatch(newAction); // ok and compiler inferred 'const doneAction: MyAction;'
const doneAction: MyAction = myStore.dispatch(newAction); // ok
const doneAction: MyAction = myStore.dispatch<MyAction>(newAction); // ok

const otherAction: OtherAction = …;
const otherDoneAction: MyAction = myStore.dispatch(otherAction); // compile error
const otherDoneAction: OtherAction = myStore.dispatch<MyAction>(otherAction); // compile error
```

Above one can be strict if parametrised by final user or almost as elastic as `(a: any) => any` without type parameters passed. So this is a win-win :)